### PR TITLE
[Release/3.1] Fix race condition issues between SinglePhaseCommit and TransactionEnded events

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Ports [dotnet/sqlclient#1042](https://github.com/dotnet/SqlClient/pull/1042) to fix below issue in System.Data.SqlClient:
- [dotnet/sqlclient#729](https://github.com/dotnet/SqlClient/issues/729) System.Data.SqlClient 4.8.2 throws TransactionAbortedException (the connection has been broken)
- Premier customer support: [Devdiv Ticket 1245724 ](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1245724 )

### Summary
A race condition exists between "Single Phase Commit" and "Transaction Ended" as both are triggered externally by delegated SqlTransaction. With new changes to doom connection in "Transaction Ended" Event (https://github.com/dotnet/corefx/pull/42937), "Commit" started failing intermittently leading to this issue. Dooming connection is essential to prevent connection re-use that leads to security issues, so that PR is valid and important.

But as a consequence, "Commit"'s inconsistent locking leads to this problem. Locking is essential in this part of code, but in "Single Phase Commit" implementation, late and split locking causes issues between Commit and Abort event handling, leading to intermittent "Transaction Aborted Exception".

This change in lock scope fixes the issue. It wasn't easily reproducible in Microsoft.Data.SqlClient but happens very often with System.Data.SqlClient 4.8.2 due to slow performance. Making test-case more rigorous and forcing latency while debugging aided in reproducing this issue.

### Customer Impact
❗ **Critical**: Premier customer application is impacted due to this issue _(followed via Emails)_.
### Regression?
**Yes**: Issue started occurring since System.Data.SqlClient v4.8.2 (PR #42937)

### Testing
It's not possible to add test case for this behavior since it's a race condition scenario and is greatly influenced by MS DTC that manages delegated transaction completion.

### Risk
**Low**: This PR does not introduce any critical design changes, but only makes changes in lock scope.
Issue has been verified by customers as well as with the repro apps available in issue [#729](https://github.com/dotnet/SqlClient/issues/729)

cc: @danmoseley @saurabh500 @David-Engel

### CODE REVIEW

**Review by JRahnama — APPROVED** [review]

**Review by David-Engel — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]

**Comment by aik-jahoda** [comment]
We have open branch, please mark the PR with Servicing-approved tag once approved.

**Pull request merged**
